### ISSUE TITLE
Designing the strategy trading interface

### ISSUE THREAD

**Issue by adamlawrencium** [adamalawrence/Ostia#11]
Now that we have a basic overview of architecture, we need to design how the strategies are actually to trade.

I propose each strategy takes in a config file:

``` javascript
config = {
  strategyName: "EMA",       // exponential moving average
  exchanges:    ["GDAX"],    // exchanges strategy will trade
  pair:         "BTC_ETC",   // or "none" for every pair
  capital:      3000,        // starting capital
  time_frame:   4,           // how long to trade
  blah:         "other param"
}
```

This creates standardization across all of our strategies and let's others design their own easily as well.
Next we need to initialize the live data sources for the strategy:

``` javascript
var exchange = config.exchanges[0];
var path = "./js/data-parsing/" + exchange + ".js";
var exchange = require(path)(allExchangeData.exchange)

var TradingHub = new TradingHub(config.exchanges[0]);
var DataHub = TradingHub.createOrderbooks();
var orderbook = DataHub.getOrderbooks();
```

These set up all the data the strategy will ever need to run.
Now that everything it set up, the strategy can run through the data and start its own trading logic . We can decide on a standard for trading logic organization, but for now:
[pseudocode]

``` javascript
function SimpleMovingAverage(orderbooks) {
  // find opportunity
  // import indicators like X-day average
  // create an order object
  var order = {
    pair:   config.pair
    buy:    true,
    sell:   false,
    amount: 100
  }
  return order;
}

// Main 
function main() {
  var start_time = getDate();
  while (start_time - getData() <  time_frame) {    // trade for specified time frame
    if (new_data == true) {
     var order = SimpleMovingAverage(orderbook);
      if (order.buy == true || order.sell == true) {    // see if there's actionable order
        TradingHub.trade(order);    // trade
      }
    }
  }
}
main();
exit(1);
```

**Comment by toshipiazza** [comment]
Talked with Adam offline; as a mentor, I lightly suggested the following:
1. Try making the configuration sufficiently general, so that the user will never have to touch the main function. Potentially it would be possible to modify the configuration, and the API main function would make use of this configuration to change its own behaviour.
   
   ``` .js
   config = {
     strategyFunc: SimpleMovingAverage,
     backlog: 1, // -1 for infinite backlog, 1 for last order
                 // useful for implementing moving average vs correct average?
     ...,
     data: {
       // general data needed for operation of the algorithm
     }
   }
   ```
2. Going off of point (1), maybe `API_main()` should take the configuration as an argument, so that the user will really never have to modify it -- this makes automation easier, but puts burden on the config to make it general enough to implement all the strategies you need.
   
   ``` .js
   function API_main(config) {
     while (...) {
       var order = config.strategyFunc(config, orderbooks);
       if (order.buy == true || order.sell == true)
         TradingHub.trade(order);
     }
   }
   ```

This is a very useful pattern, since end users will not want to muck around with logic related to timing, etc. Also, the less of the whole API the user is able to access, the better! We don't necessarily want the user to modify the TradingHub(), because if they use it incorrectly I'm assuming it could break all of our metrics, etc.

This also gives us the leeway to change the operation of the main function, since we could either implement it as a poll like we are appearing to do now, or we could implement it as a callback on DataHub() to wake up when data becomes available, etc.

**Comment by toshipiazza** [comment]
Another thought -- if we want a way for the user to submit code, perhaps we can consider using node's module system, i.e.

``` .js
// client.js
function superComplexTradingAlgorithm(config, dataHub) {
  return { buy: true, sell: false }; // always buy 
}
var config = { strategyFunc: superComplexTradingAlgorithm, ... };
module.exports = config;
```

Here, we're **requiring** that the user submit code in client.js. That way, we can write a simple driver like so:

``` .js
// api_main.js
var client_config = require("client");
API_main(client_config);
```

**Comment by Mingling94** [comment]
1. I see and agree with what @toshipiazza is saying. The config needs to be durable enough to represent any possible trading strategy. This is easy since the config could have member functions that perform logic, as toshi alludes to with:
   `var order = config.strategyFunc(config, orderbooks);`
2. This is as opposed to @adamalawrence's suggestion of having the config store a string/enum/some sort of flag that, in the main function, affects the logic being ran. 

Approach #2 introduces [tight coupling](http://stackoverflow.com/questions/2832017/what-is-the-difference-between-loose-coupling-and-tight-coupling-in-object-orien) (bad). Approach #1 supports [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) (good). Toshi's suggestion gets my vote.

BTW, we're making this out to be a website that people make accounts on, input their API keys, and let our servers handle their trading, right? If we're planning for a client-server model, most of our code right now is server-only. The client ought to have a responsive settings menu that makes API calls to change the config file.

**Comment by Mingling94** [comment]
@adamalawrence good github practice necessitates that you provide a reason when you close issues. Like this:

**Comment by Mingling94** [comment]
Implemented in #13

**Comment by adamlawrencium** [comment]
Got it. I also archived the post in our Wiki.
